controller/konnect/reconciler_generic.go

```diff
@@ -0,0 +1,263 @@
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				KonnectEntityAPIAuthConfigurationRefValidConditionType,
+				metav1.ConditionFalse,
+				KonnectEntityAPIAuthConfigurationRefReasonInvalid,
+				fmt.Sprintf("referenced KonnectAPIAuthConfiguration %s is invalid", apiAuthRef),
+				ent.GetGeneration(),
+			),
+			ent,
+		)
+
+		return ctrl.Result{}, nil
+	}
+
+	// If the referenced APIAuthConfiguration is valid, set the "APIAuthRefValid"
+	// condition to true.
+	k8sutils.SetCondition(
+		k8sutils.NewConditionWithGeneration(
+			KonnectEntityAPIAuthConfigurationRefValidConditionType,
+			metav1.ConditionTrue,
+			KonnectEntityAPIAuthConfigurationRefReasonValid,
+			fmt.Sprintf("referenced KonnectAPIAuthConfiguration %s is valid", apiAuthRef),
+			ent.GetGeneration(),
+		),
+		ent,
+	)
+	if err := r.Client.Status().Update(ctx, ent); err != nil {
+		if k8serrors.IsConflict(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to update status with %s condition: %w", KonnectEntityAPIAuthConfigurationRefValidConditionType, err)
+	}
+
+	// NOTE: We need to create a new SDK instance for each reconciliation
+	// because the token is retrieved in runtime through KonnectAPIAuthConfiguration.
+	sdk := sdkkonnectgo.New(
+		sdkkonnectgo.WithSecurity(
+			sdkkonnectgocomp.Security{
+				PersonalAccessToken: sdkkonnectgo.String(apiAuth.Spec.Token),
+			},
+		),
+		sdkkonnectgo.WithServerURL("https://"+apiAuth.Spec.ServerURL),
+	)
+
+	if delTimestamp := ent.GetDeletionTimestamp(); !delTimestamp.IsZero() {
+		logger.Info("resource is being deleted")
+		// wait for termination grace period before cleaning up
+		if delTimestamp.After(time.Now()) {
+			logger.Info("resource still under grace period, requeueing")
+			return ctrl.Result{
+				// Requeue when grace period expires.
+				// If deletion timestamp is changed,
+				// the update will trigger another round of reconciliation.
+				// so we do not consider updates of deletion timestamp here.
+				RequeueAfter: time.Until(delTimestamp.Time),
+			}, nil
+		}
+
+		if controllerutil.RemoveFinalizer(ent, KonnectCleanupFinalizer) {
+			if err := Delete[T, TEnt](ctx, sdk, logger, r.Client, ent); err != nil {
+				return ctrl.Result{}, err
+			}
+			if err := r.Client.Update(ctx, ent); err != nil {
+				if k8serrors.IsConflict(err) {
+					return ctrl.Result{Requeue: true}, nil
+				}
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s: %w", KonnectCleanupFinalizer, err)
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: relying on status ID is OK but we need to rethink this because
+	// we're using a cached client so after creating the resource on Konnect it might
+	// happen that we've just created the resource but the status ID is not there yet.
+	//
+	// We should look at the "expectations" for this:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go
```

Comment:
I think this is fine. We are returning after setting the ID, and the status update is going to re-trigger another reconciliation loop. In the new loop, we are guaranteed the status has been properly updated, as that specific update is the one that triggered the new loop. (Unless other requests are already in the queue maybe?)

Reply:
Imagine 2 subsequent reconciliations. The first one creates a Konnect resource because the Konnect ID is empty. The second one kicks in: the status should be set but since we're working with a cached client it might not be there yet. This would cause lines below to attempt creating the entity on Konnect again.

That's what this comment tries to cover.

Reply:
yes, so this is the case I mentioned with 
> Unless other requests are already in the queue maybe?
?

It should be a harmless problem, as the 2nd creation fails and an error is returned. Bad UX though

Reply:
> Bad UX though

Yes. Additionally this will set the status accordingly, for a brief while, but still.

For this reason I'd prefer to leave the comment in. Does that sound good?